### ISSUE TITLE
Access to loss and its gradient

### ISSUE THREAD

**Issue by fujiisoup** [blei-lab/edward#708]
(from #529 )

I sometimes want to add a regularization term to the loss or clip its gradient to stabilize the learning process in particular for variational models with neural network.
Is there any workaround about it?
Regularization has been proposed several times (#529 and #531) but seems not to be merged yet.

I think it would be great if we could access to the loss or its gradient in very general way.
My current proposal is to add additional argument in `VariationalInference.initialize`,

```python
def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
                 gradient_generator=gradient_generator_reg,
                 global_step=None, *args, **kwargs):
```
so that we can pass a callable related to the regularization or gradient clipper, 
such as
```python
def gradient_generator_reg(loss, grad, v):
  loss = loss + reg * tf.square(v)
  return loss, tf.gradient(loss, v)
```

Any thoughts are welcome.

**Comment by davharris** [comment]
More access to the loss function would be a big help for my work too. Even if the user doesn't have as much control as suggested in the example above, being able to say `inference.loss += f(z)` would be a big help.  This would be analogous to `target +=` statements in Stan (formerly called `increment_log_prob`). 

In Stan, there are models that are difficult or impossible to write without incrementing the target (negative loss). Edward has some great features that help mitigate this need (e.g. automatic transformations that account for the Jacobian) but it looks like there have still been a number of cases where this would still a very useful addition.

**Comment by davharris** [comment]
I've posted a (somewhat indirect) way to add to the loss function here: https://github.com/blei-lab/edward/issues/529#issuecomment-341943910

**Comment by fujiisoup** [comment]
@davharris
Thanks for the information. It would help!

FYI, I am using the following workaround for the gradient clipping.

```Python
class ClippedAdamOptimizer(tf.train.AdamOptimizer):
    """
    Clipped version adam optimizer, where its gradient is clipped by value
    so that it cannot be too large.
    """
    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999,
                 epsilon=1e-08, use_locking=False,
                 clip_func=lambda x: tf.clip_by_value(x, 1.0e-4, 1.0e4),
                 name='Adam'):
        super(ClippedAdamOptimizer, self).__init__(
            learning_rate, beta1, beta2, epsilon, use_locking, name)
        self._clip_func = clip_func

    def compute_gradients(self, loss, var_list=None, gate_gradients=GATE_OP,
                          aggregation_method=None,
                          colocate_gradients_with_ops=False, grad_loss=None):
        grad_and_vars = super(ClippedAdamOptimizer, self).compute_gradients(
            loss, var_list, gate_gradients, aggregation_method,
            colocate_gradients_with_ops, grad_loss)
        # clip func
        if self._clip_func is None:
            return grad_and_vars
        return [(self._clip_func(g, v) if g is not None else (g, v)
                for g, v in grad_and_vars)]
```

**Comment by JohnReid** [comment]
That's very useful thanks. Note to self: `GATE_OP` is `tf.train.Optimizer.GATE_OP`
